tests/test_controllers.py

```diff
@@ -0,0 +1,165 @@
+                        break
+                    start_idx += robot.controllers[c].command_dim
+                forward_action[start_idx] = 0.1
+                side_action[start_idx + 1] = 0.1
+                up_action[start_idx + 2] = 0.1
+            actions["zero"][robot.name] = zero_action
+            actions["forward"][robot.name] = forward_action
+            actions["side"][robot.name] = side_action
+            actions["up"][robot.name] = up_action
+
+            # Add base movement action if locomotion robot
+            base_move_action = th.zeros(robot.action_dim)
+            if isinstance(robot, LocomotionRobot):
+                c_name = "base"
+                start_idx = 0
+                for c in robot.controller_order:
+                    if c == c_name:
+                        break
+                    start_idx += robot.controllers[c].command_dim
+                base_move_action[start_idx] = 0.1
+            actions["base_move"][robot.name] = base_move_action
+
+        # Take 5 steps
+        for _ in range(5):
+            env.step(actions["zero"])
+
+        # Update initial state
+        env.scene.update_initial_state()
+
+        # For each action set, reset all robots, then run actions and see if arm moves in expected way
+        for action_name, action in actions.items():
+
+            # Reset env
+            env.reset()
+
+            # Record initial poses
+            initial_eef_pose = dict()
+            for i, robot in enumerate(env.robots):
+                initial_eef_pose[robot.name] = {arm: robot.get_relative_eef_pose(arm=arm) for arm in robot.arm_names}
+
+            # Take 10 steps with given action and check for error
+            for _ in range(n_steps[action_name]):
+                env.step(action)
+
+            for i, robot in enumerate(env.robots):
+                for arm in robot.arm_names:
+                    init_pos, init_quat = initial_eef_pose[robot.name][arm]
+                    curr_pos, curr_quat = robot.get_relative_eef_pose(arm=arm)
+                    is_valid_pos = pos_err_checks[action_name](curr_pos, init_pos)
+                    assert is_valid_pos, (
+                        f"Got mismatch for controller [{controller}], robot [{robot.model_name}], action [{action_name}]\n"
+                        f"curr_pos: {curr_pos}, init_pos: {init_pos}"
+                    )
+                    ori_err_normalized = th.norm(
+                        T.quat2axisangle(T.mat2quat(T.quat2mat(init_quat).T @ T.quat2mat(curr_quat)))
+                    ).item() / (np.pi * 2)
+                    ori_err = np.abs(np.pi * 2 * (np.round(ori_err_normalized) - ori_err_normalized))
+                    assert ori_err < 0.1, (
+                        f"Got mismatch for controller [{controller}], robot [{robot.model_name}], action [{action_name}]\n"
+                        f"curr_quat: {curr_quat}, init_quat: {init_quat}, err: {ori_err}"
+                    )
```

Comment:
can we do a quantitative test for orientation as well? e.g. command the eef to rotate 45 degrees and confirm it's the case.

Reply:
done!